// Copyright 2024 Flower Labs GmbH. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

syntax = "proto3";

package flwr.proto;

message DoubleList { repeated double vals = 1; }
message SintList { repeated sint64 vals = 1; }
message UintList { repeated uint64 vals = 1; }
message BoolList { repeated bool vals = 1; }
message StringList { repeated string vals = 1; }
message BytesList { repeated bytes vals = 1; }

message Array {
  string dtype = 1;
  repeated int32 shape = 2;
  string stype = 3;
  bytes data = 4;
}

message MetricRecordValue {
  oneof value {
    // Single element
    double double = 1;
    sint64 sint64 = 2;
    uint64 uint64 = 3;

    // List types
    DoubleList double_list = 21;
    SintList sint_list = 22;
    UintList uint_list = 23;
  }
}

message ConfigRecordValue {
  oneof value {
    // Single element
    double double = 1;
    sint64 sint64 = 2;
    uint64 uint64 = 3;
    bool bool = 4;
    string string = 5;
    bytes bytes = 6;

    // List types
    DoubleList double_list = 21;
    SintList sint_list = 22;
    UintList uint_list = 23;
    BoolList bool_list = 24;
    StringList string_list = 25;
    BytesList bytes_list = 26;
  }
}

message ArrayRecord {
  message Item {
    string key = 1;
    Array value = 2;
  }
  repeated Item items = 1;
}

message MetricRecord {
  message Item {
    string key = 1;
    MetricRecordValue value = 2;
  }
  repeated Item items = 1;
}

message ConfigRecord {
  message Item {
    string key = 1;
    ConfigRecordValue value = 2;
  }
  repeated Item items = 1;
}

message RecordDict {
  message Item {
    string key = 1;
    oneof value {
      ArrayRecord array_record = 2;
      MetricRecord metric_record = 3;
      ConfigRecord config_record = 4;
    }
  }
  repeated Item items = 1;
}
